sdk/iot/hub/src/az_iot_hub_client.c

```diff
@@ -47,6 +48,8 @@ AZ_NODISCARD az_result az_iot_hub_client_user_name_get(
   AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
   AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
 
+  mqtt_user_name = _az_span_set_length(mqtt_user_name, 0);
```

Comment:
This is just a comment: I would expect function to NOT set the length to 0 explicitly for the span that is passed in. Our code in Core never does this. Instead, our code just ignores the passed-in span's length.

I see calling _az_span_set_length in order to set the output parameter to the input span's ptr/cap & and the function's calculated length AFTER the operation is performed. 

If you want to explicitly set the passed-in span's length to 0, you can. But to me, this feels like code bloat and perf hit for no value.

Reply:
The problem there is I guess related to what we were discussing on Friday. If we don't set it here, the user would have to set the length to zero themselves. Otherwise all of these API's will act as "append" methods.

Take an example where someone has a span for a telemetry topic. They get it, use it, but then have some properties to append. So they add the props and go back to get the telemetry topic again, passing the same span which had the first topic. If WE set it to zero, they will get back the fresh topic starting from `az_span_ptr(tel_topic)`. If we don't, the user must set it to zero or else they will build the topic starting at `az_span_ptr(tel_topic) + first_len`

Reply:
I don't understand the change either: input and output spans are different. You shouldn't touch the first one at all. The second can be initialized with the first one's buffer and capacity and zero-len.

@JeffreyRichter we're using append so I'm not sure how that would work w/o having len=0. What APIs should we be using instead for example to copy from `client->_internal.device_id` to the output span?

Reply:
@CIPop: Your mental model & mine WERE in agreement.
However, @danewalton & I had a separate conversation and he has a great point. 

What his code does is this: He takes the input (passed by value) span and when not appending, he sets the length to 0. When appending, he leaves it as-is. 

Then he uses this span in the function's code and when the function is about to return SUCCESS, he copies the passed-in, now-mutated span into the output span. The reason he does it this way is because if he mutated the output span and the function FAILS, then then the function WILL corrupt the customer's span if the caller passed the same span by value and by-reference:
   az_result r = some_function(my_span, &my_span);  
   // If the function above fails, then Dane's way doesn't corrupt my_span. 

This has value and I think we'd adopt Dane's pattern in Core (for functions that are not doing this already).

Reply:
I've answered to an internal thread. Looking closer, this function is actually copying the struct again so we now have one more copy of the same struct on the stack:
1. From the original call to the API
2. Another copy to _az_span_set_length

2x for above 2 stack locations, one for mqtt_user_name = another at the end to copy it over _the same location_ in out_mqtt_user_name.

I really don't see the reasons to have this overhead.

My solution: take a single argument (&mqtt_user_name) and only touch len by using either something like az_clear(&span) or az_copy() with the first string then using az_span_append (or similar). If the function fails the mqtt_user_name will be guaranteed to have the same buffer and capacity (since none of the APIs used alter those). It is not guaranteed to have the same len - since that len is associated with the values within the buffer which have been destroyed by the operations anyway.

Reply:
The function is inline and so that shouldn't happen - the compiler should optimize the extra copy away. We could make other az_span_* functions inline too but then they can't have precondition checks. We're constantly at a tug of war.

Reply:
We've discussed this pattern more offline with @JeffreyRichter as well as the IoT team and the current thinking is to potentially move to `char*, capacity, &len` instead.

That said, I still don't think there is a need to set the len: @danewalton, @JeffreyRichter couldn't we just use `az_span_copy` for the first operation? That would reset the len.

Reply:
We can't use `az_span_copy` because that actually copies the buffer over to another buffer: 
https://github.com/Azure/azure-sdk-for-c/blob/88baf33becd27795facc78457ef0e2f070acbf96/sdk/core/core/src/az_span.c#L242-L245

By setting it to `0`, I am in effect ignoring the length. I am using a span already on the stack, provided by the user, and the only thing I want to change is the length so that the first call to `append` I make is at `ptr` and not the passed length (`ptr + length` which in that case would act as an append).

Not setting it to 0 causes it to have the same functionality as it did originally, which is what you brought up as wrong.

Reply:
I don't understand: isn't

```C
az_span_copy(mqtt_user_name, client->_internal.iot_hub_hostname, &mqtt_user_name));
```

equivalent to 

```C
mqtt_user_name = _az_span_set_length(mqtt_user_name, 0);
az_span_append(mqtt_user_name, client->_internal.iot_hub_hostname, &mqtt_user_name));
```
?

Reply:
Ahhhh I see what you’re saying sorry I misunderstood. I thought you meant replace the `set_length` with the `copy` and assumed you were intending to copy the span struct values not the buffer values. No I think that would work you are right.